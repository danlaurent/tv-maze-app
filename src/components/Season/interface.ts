import { IEpisode } from '../../interface/shows';

export interface ISeason {
  season: {
    season: number;
    episodes: IEpisode[];
  };
  onEpisodePress?: (episode: IEpisode) => void;
  testID?: string;
}
