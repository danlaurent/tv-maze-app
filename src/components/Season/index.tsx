import React from 'react';
import { View, Text, StyleSheet, TouchableOpacity } from 'react-native';
import { FontAwesome5 } from '@expo/vector-icons';
import { ISeason } from './interface';
import {
  DARK_QUANTUM_BLUE,
  GRAY,
  LIGHT_GRAY,
  QUANTUM_BLUE,
} from '../../constants/styles/colors';

const styles = StyleSheet.create({
  header: {
    height: 56,
    paddingHorizontal: 12,
    paddingVertical: 4,
    justifyContent: 'center',
    borderBottomWidth: 1,
    borderTopWidth: 1,
    borderColor: DARK_QUANTUM_BLUE,
    backgroundColor: QUANTUM_BLUE,
  },
  headerText: {
    fontSize: 16,
    fontWeight: '700',
    color: LIGHT_GRAY,
  },
  episodesContainer: {
    backgroundColor: DARK_QUANTUM_BLUE,
  },
  episode: {
    height: 48,
    paddingHorizontal: 20,
    paddingVertical: 4,
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'space-between',
    borderBottomWidth: 1,
    borderColor: QUANTUM_BLUE,
  },
  episodeText: {
    fontSize: 16,
    fontWeight: '700',
    color: GRAY,
  },
});

const Season = ({ season, onEpisodePress = () => {}, testID }: ISeason) => (
  <View testID={testID}>
    <View testID={`${testID}_header`} style={styles.header}>
      <Text testID={`${testID}_headerText`} style={styles.headerText}>
        Season {season.season}
      </Text>
    </View>
    <View style={styles.episodesContainer}>
      {season.episodes.map((episode) => (
        <TouchableOpacity
          testID={`${testID}_episodeButton_${episode.id}`}
          style={styles.episode}
          key={episode.id}
          onPress={() => onEpisodePress(episode)}
        >
          <Text
            testID={`${testID}_episodeText_${episode.id}`}
            style={styles.episodeText}
            numberOfLines={1}
          >
            {episode.number} - {episode.name}
          </Text>
          <FontAwesome5 name='chevron-right' size={20} color={GRAY} />
        </TouchableOpacity>
      ))}
    </View>
  </View>
);

export default Season;
