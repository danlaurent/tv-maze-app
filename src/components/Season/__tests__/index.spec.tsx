import React from 'react';
import { render, RenderAPI, fireEvent } from '@testing-library/react-native';
import Season from '..';

const onEpisodePressMock = jest.fn();

const seasonMock = {
  season: 1,
  episodes: [
    {
      id: '1',
      name: 'Test Episode',
      number: 1,
      season: 1,
      image: {
        medium: '',
        original: '',
      },
      summary: 'A test episode',
      rating: {
        average: 10,
      },
    },
  ],
};

const emptySeasonMock = {
  season: 2,
  episodes: [],
};

describe('Season', () => {
  let tree: RenderAPI;

  beforeEach(() => {
    tree = render(
      <Season
        testID='testSeason'
        season={seasonMock}
        onEpisodePress={() => onEpisodePressMock()}
      />
    );
  });
  afterEach(() => {
    jest.clearAllMocks();
  });

  it('renders correctly', () => {
    const snapshot = tree.toJSON();
    expect(snapshot).toMatchSnapshot();
  });

  it('shows the Season on header', () => {
    expect(tree.getByText('Season 1')).toBeDefined();
  });

  it('shows the episode number and name', () => {
    expect(tree.getByText('1 - Test Episode')).toBeDefined();
  });

  describe('When the episode is pressed', () => {
    beforeEach(async () => {
      await fireEvent.press(tree.getByTestId('testSeason_episodeButton_1'));
    });

    it('fires onEpisodePress', () => {
      expect(onEpisodePressMock).toBeCalled();
    });
  });

  describe('If there is no episodes', () => {
    beforeEach(() => {
      tree = render(<Season testID='testSeason' season={emptySeasonMock} />);
    });

    it('do not show the episode button', () => {
      expect(tree.queryByTestId('testSeason_episodeButton_1')).toBeNull();
    });

    it('do not show the episode name and number', () => {
      expect(tree.queryByTestId('testSeason_episodeText_1')).toBeNull();
    });
  });
});
