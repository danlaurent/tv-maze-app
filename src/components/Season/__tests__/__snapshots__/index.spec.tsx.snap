// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Season renders correctly 1`] = `
<View
  testID="testSeason"
>
  <View
    style={
      Object {
        "backgroundColor": "#121322",
        "borderBottomWidth": 1,
        "borderColor": "#0c0d18",
        "borderTopWidth": 1,
        "height": 56,
        "justifyContent": "center",
        "paddingHorizontal": 12,
        "paddingVertical": 4,
      }
    }
    testID="testSeason_header"
  >
    <Text
      style={
        Object {
          "color": "#c6c7d2",
          "fontSize": 16,
          "fontWeight": "700",
        }
      }
      testID="testSeason_headerText"
    >
      Season 
      1
    </Text>
  </View>
  <View
    style={
      Object {
        "backgroundColor": "#0c0d18",
      }
    }
  >
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "borderBottomWidth": 1,
          "borderColor": "#121322",
          "flexDirection": "row",
          "height": 48,
          "justifyContent": "space-between",
          "opacity": 1,
          "paddingHorizontal": 20,
          "paddingVertical": 4,
        }
      }
      testID="testSeason_episodeButton_1"
    >
      <Text
        numberOfLines={1}
        style={
          Object {
            "color": "#6d6f7e",
            "fontSize": 16,
            "fontWeight": "700",
          }
        }
        testID="testSeason_episodeText_1"
      >
        1
         - 
        Test Episode
      </Text>
      <Text />
    </View>
  </View>
</View>
`;
